src/rust/node-identifier/src/node_identifier.rs

```diff
@@ -0,0 +1,234 @@
+                            )
+                        )
+                        .map(|identified_to_edge| (identified_to_edge, &edge.edge_name))
+                });
+
+            // add all identified edges into the `identified_graph`
+            for (identified_to_key, edge_name) in identified_edges {
+                identified_graph.add_edge(
+                    edge_name.to_owned(),
+                    identified_from_key.to_owned(),
+                    identified_to_key.to_owned(),
+                );
+            }
+        }
+    }
+
+    #[tracing::instrument(skip(self, unidentified_subgraph))]
+    pub(crate) async fn handle_event(
+        &self,
+        unidentified_subgraph: GraphDescription,
+    ) -> Result<IdentifiedGraph, Result<(IdentifiedGraph, NodeIdentifierError), NodeIdentifierError>>
+    {
+        /*
+           1. for the nodes in the unidentified graph, do the following
+               1. identify each node
+               2. record the old node key -> new node key mapping
+               3. record which nodes failed to identify
+               4. place the correctly identified nodes into a new, identified graph
+           2. for the edges in the unidentified graph, do the following
+               1. for both the 'from' and 'to' sections, look up the old-node-key -> new-node-key map to fetch updated values
+                   1. if both are present, update edge and add it to the identified graph
+                   2. if one or more are missing, warn with an appropriate error message and skip the edge
+           3. if identified graph is empty, return as full error
+           4. if any nodes failed to identify or any errors occurred, return as a partial error
+           5. return as full graph
+        */
+
+        tracing::info!(
+            message="Identifying a new graph",
+            node_count=?unidentified_subgraph.nodes.len(),
+            edge_count=?unidentified_subgraph.edges.len(),
+        );
+
+        if unidentified_subgraph.is_empty() {
+            tracing::warn!("Received empty subgraph.");
+            return Err(Err(NodeIdentifierError::EmptyGraph));
+        }
+
+        let mut identified_graph = IdentifiedGraph::new();
+
+        let (identified_nodekey_map, attribution_failure) = self
+            .identify_nodes(&unidentified_subgraph, &mut identified_graph)
+            .await;
+
+        tracing::info!(
+            message="Performed node identification",
+            total_edges=?unidentified_subgraph.nodes.len(),
+            identified_edges=?identified_graph.nodes.len()
+        );
+
+        self.identify_edges(
+            &unidentified_subgraph,
+            &mut identified_graph,
+            identified_nodekey_map,
+        );
+
+        tracing::info!(
+            message = "Performed edge identification",
+            total_edges = unidentified_subgraph.edges.len(),
+            identified_edges = identified_graph.edges.len()
+        );
```

Comment:
why are some of these traces `=?` and some ` = ` ? I believe that calls Debug::fmt or something